pkg/gadgets/socket-collector/collector.go

```diff
@@ -0,0 +1,26 @@
+// +build !docs
```

Comment:
What's the purpose of this `docs` tag? To avoid the dependency on bpf2go when generating the docs?

Reply:
Yes, the new skeleton embeds the BPF binary file instead of adding it as a string like this:

```
//go:embed itertcpv4_bpfel.o
var _IterTCPv4Bytes []byte
```

Having the file there was causing problems for docs generation so I removed the dependency of the "github.com/kinvolk/inspektor-gadget/pkg/gadgets/socket-collector/tracer" from the gadget using a wrapper function `runCollector` defined twice: (1) For docs: Does not require the tracer and (2) For build: Which is the actual implementation.

The doc generation problem is gone now but I am still having some problems with the test part. I am working on it.

BTW, I will add a proper message in the commit message and PR description once I find a final solution.

Reply:
I'll be happy to know more details about that error. I think having this building flags everywhere increases complexity and creates confusion.

Reply:
Sure, let me try ... The situation is that operations like docs generation and unit-tests execution are failing because they are being requested to embed BPF binary files that were not generated. The problem is that it is correct that those binary files were not generated because these operations do not really use them. Therefore, we have a sort of "deadlock":

```
go run -tags docs cmd/gen-doc/gen-doc.go -repo /home/runner/work/inspektor-gadget/inspektor-gadget 
Error: pkg/gadgets/socket-collector/tracer/itertcpv4_bpfel.go:113:12: pattern itertcpv4_bpfel.o: no matching files found 
```

```
go test -test.v ./... 
Error: pkg/gadgets/socket-collector/tracer/itertcpv4_bpfel.go:113:12: pattern itertcpv4_bpfel.o: no matching files found 
Error: pkg/gadgets/socket-collector/tracer/itertcpv4_bpfel.go:113:12: pattern itertcpv4_bpfel.o: no matching files found 
Error: make: *** [Makefile:78: test] Error 1 
Error: Process completed with exit code 2.
```

So far, we have solved this problem by using the `-tags` flag and the build constraints ([Ref](https://pkg.go.dev/cmd/go#hdr-Build_constraints)). This method allows us to use the source files where we are embedding the BPF binary files only during `gadget-container` building (tag `withebpf`). Furthermore, using the tag `docs`, we avoided the need for `libseccomp` during the doc generation.

What is clear enough?

Reply:
I agree with you that it increases the complexity so it would be great if we can find a simpler solution.

Reply:
I understood the problem. Is it that bad to embed those files also when running the tests or creating the docs?

Reply:
The problem is that those files do not exist when running the unit-tests or generating the docs. I see a couple of options to have them in place when needed:

1. **Add the generation of those binary files as pre-requirement in the Makefile**: If we do so, we are requesting the user/developer to have LLVM, Clang and kernel header to just generate the documentation or run unit-tests that do not even test the BPF code.
2. **Option (1) but running the doc generation and unit-test on the docker builder image we use to compile the container-gadget**: We are not adding requirements anymore to the user/developer but it would increase considerably the time to generate the documentation or run the unit-test.
3. **Ship the binary files with the repo**: At least for what concerns the socket-collector, the binary was already being shipped but as a hex string inside the skeleton. Therefore, including the `.o` will not change that much. 

In my opinion, at least for this PR, option (3) seems to be the best compromise. However, in general, it does not seem that good to include a set of binary files in the repo that will not be actually used but just to pass the golang files check. I mean, we could even create empty `.o` files and the doc generation and the unit-test will succeed (Should we consider this as an option? :thinking:).

What do you all think?

cc\ @marga-kinvolk

Reply:
I think I still prefer to fix this with tags, if possible, although it is not ideal.

Reply:
I agree that tags are the easiest way now. Other solutions seem too complicated to me. Thanks for the detailed explanation about the issue.

Reply:
I found a way to bypass both errors (gen docs and unit-tests) using the `withebpf` tag we already had. PTAL.

In any case, I opened an issue on cilium/ebpf to report this situation, share how we are solving it and ask for suggestions.